### ISSUE TITLE
Dynamically Provisioned AWS EBS Volumes Fail to Delete

### ISSUE THREAD

**Issue by FrederikNS** [kubernetes/kubernetes#40220]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): persistent volume ebs delete failed

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`): 1.4.6

**Environment**:
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): Debian GNU/Linux 8 (jessie)
- **Kernel** (e.g. `uname -a`): Linux ip-172-20-65-129 4.4.26-k8s #1 SMP Fri Oct 21 05:21:13 UTC 2016 x86_64 GNU/Linux
- **Install tools**: kops
- **Others**:

**What happened**:
I created a StorageClass for AWS EBS volumes, then I created a PersistentVolumeClaim matching it. Afterwards I deleted the PersistentVolumeClaim. The PersistentVolumeClaim disappeared, but the PersistentVolume still exists, and changed status to Failed. The message reads:

    Delete of volume "pvc-302d9601-df15-11e6-a28f-062d31f06f1f" failed: error deleting EBS volumes: VolumeInUse: Volume vol-079a9bccfff60e463 is currently attached to i-00bfbade5895a8930

It seems that Kubernetes are trying to delete the volume without detaching the volume from the node.

**What you expected to happen**:
I expected Kubernetes to delete the PersistentVolume.

**How to reproduce it** (as minimally and precisely as possible):
Create a cluster on AWS.
Add StorageClass:

    kind: StorageClass
    apiVersion: storage.k8s.io/v1beta1
    metadata:
      name: ssd
      annotations:
        storageclass.beta.kubernetes.io/is-default-class: "true"
    provisioner: kubernetes.io/aws-ebs
    parameters:
      type: gp2

Add a PersistentVolumeClaim:

    kind: PersistentVolumeClaim
    apiVersion: v1
    metadata:
      name: elasticsearch-data
      namespace: logging
      annotations:
        volume.beta.kubernetes.io/storage-class: "ssd"
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 10Gi

Delete the PersistentVolumeClaim.
Watch the PersistentVolume status.

**Anything else do we need to know**:
I tried to "force detach" the volume in the AWS Console, and after a short while Kubernetes deleted the volume on it's own.

I'll be happy to provide any additional info you need.

**Comment by gnufied** [comment]
Is the PV being used from a pod? If yes - you won't be able to delete it.

**Comment by FrederikNS** [comment]
I did use the PV in a pod, but I also deleted the pod, and the PV remains

**Comment by gnufied** [comment]
Can you list exact steps to reproduce the bug? i.e - order in which you created and deleted pods and pvcs.

**Comment by jingxu97** [comment]
@FrederikNS could you please list the steps which also includes create and delete pod? Thanks!

**Comment by FrederikNS** [comment]
I use the following 4 yaml configs:
[elasticsearch.yaml.txt](https://github.com/kubernetes/kubernetes/files/723582/elasticsearch.yaml.txt)
[elasticsearch-volume.yaml.txt](https://github.com/kubernetes/kubernetes/files/723581/elasticsearch-volume.yaml.txt)
[logging-namespace.yaml.txt](https://github.com/kubernetes/kubernetes/files/723584/logging-namespace.yaml.txt)
[ssd-storage-class.yaml.txt](https://github.com/kubernetes/kubernetes/files/723583/ssd-storage-class.yaml.txt)

Then i follow this procedure:
1. `kubectl create -f ssh-storage-class.yaml`
2. `kubectl create -f logging-namespace.yaml`
3. `kubectl create -f elasticsearch-volume.yaml`
4. Ensure that volume is in status `Bound`
5. `kubectl create -f elasticsearch.yaml`
6. Wait for elasticsearch to be ready.
7. `kubectl delete deployment elasticsearch`
8. Wait for the pod to be completely terminated.
9. `kubectl delete pvc elasticsearch-data`
10. The Persistent Volume Claim is now gone, but the Persistent Volume still exists and has changed status to `Failed`

The Persistent Volume's description now reads:

    Name:		pvc-b4b48f0b-e151-11e6-a28f-062d31f06f1f
    Labels:		failure-domain.beta.kubernetes.io/region=eu-west-1
    		failure-domain.beta.kubernetes.io/zone=eu-west-1a
    Status:		Failed
    Claim:		logging/elasticsearch-data
    Reclaim Policy:	Delete
    Access Modes:	RWO
    Capacity:	300Gi
    Message:	Delete of volume "pvc-b4b48f0b-e151-11e6-a28f-062d31f06f1f" failed: error deleting EBS volumes: VolumeInUse: Volume vol-004c26e60a81f226c is currently attached to i-08c05f85f55eabe44
    		status code: 400, request id: 
    Source:
        Type:	AWSElasticBlockStore (a Persistent Disk resource in AWS)
        VolumeID:	aws://eu-west-1a/vol-004c26e60a81f226c
        FSType:	ext4
        Partition:	0
        ReadOnly:	false
    Events:
      FirstSeen	LastSeen	Count	From				SubobjectPath	Type		Reason			Message
      ---------	--------	-----	----				-------------	--------	------			-------
      7m		7m		1	{persistentvolume-controller }			Warning		VolumeFailedDelete	Delete of volume "pvc-b4b48f0b-e151-11e6-a28f-062d31f06f1f" failed: error deleting EBS volumes: VolumeInUse: Volume vol-004c26e60a81f226c is currently attached to i-08c05f85f55eabe44
    		status code:

I have seen it stay in that state for 4 hours, but if I go the the AWS Console, and `force detach` the specific volume, then Kubernetes deletes the volume within a few seconds.

**Comment by jingxu97** [comment]
@FrederikNS can I ask whether you could reproduce the issue? It seems like volume controller fail to detach the volume, but we need kube-controller-manager log to confirm that.

**Comment by FrederikNS** [comment]
I also bumped into an issue yesterday where volumes would never actually mount. Unfortunately I didn't have the time, so I tried to upgrade the cluster from 1.4.6 to 1.4.7, and now both problems have disappeared. Might be the upgrade that fixed it, or it might just be the fresh machines that fixed it...

Unfortunately the logs would have been wiped in the upgrade, so I can't really help that much now... If the problem reappears, I'll definitely post the kube-controller-manager logs.

Feel free to close the issue.
